README

oz_holidays - Ruby
------------------

gem install oz_holidays

require 'oz_holidays'

> OzHolidays.national.keys.grep /^2012/

=> ["2012-01-02", "2012-01-26", "2012-04-06", "2012-04-09", "2012-04-25",
    "2012-12-25", "2012-12-26"] 


> OzHolidays['NSW']['2013-04-25']

=> "ANZAC Day" 




Javascript / JSON
-----------------

See javascript/ directory or run OzHolidays.to_json



Contact the author
------------------

Andrew Snow <andrew@modulus.org>
Andys^ on irc.freenode.net


CHANGELOG
v1.2014. Update for 2014
v1.2013. Update with 2013 holidays
v1.0. National & State Holidays for 2011 and 2012



test/test_all.rb
require 'test/unit'
require "#{File.dirname(__FILE__)}/../lib/oz_holidays"
require 'date'
require 'json'


class TestOzHolidays < Test::Unit::TestCase
  def test_new_years_day
    assert_match /New Year/, OzHolidays.national[find_nyd_hol]
  end
  
  def test_anzac_day
    assert_not_nil OzHolidays['NSW'][Date.parse("#{Date.today.year}-04-25")]
  end
  
  def test_all
    assert_kind_of Array, OzHolidays.all
    assert_kind_of Hash, OzHolidays.all.first
  end
  
  def test_to_json
    assert_not_nil OzHolidays.to_json
  end
  
  def test_by_date
    assert_kind_of Hash, OzHolidays.by_date
    assert_kind_of Hash, OzHolidays.by_date.values.first
  end
  
  
private
  def find_nyd_hol
    nyd = Date.parse("#{Date.today.year}-01-01")
    nyd += 1 while(nyd.wday < 1 || nyd.wday > 5)
    nyd
  end
end


gen/parse.rb

require 'date'

state = nil
holidays11 = Hash.new {|h,k| h[k] = {} }
holidays12 = Hash.new {|h,k| h[k] = {} }

STDIN.each_line do |line|
  if line =~ /^\*(.+)$/
    state = $1.strip
  elsif line =~ /^\w+/
    dayname, day11, day12 = line.split(/\t/)
    dayname.strip!
    dayname.gsub! /[^a-zA-Z ]/, ''

    holidays11[state][Date.parse(day11 + ' 2011').to_s] = dayname
    holidays12[state][Date.parse(day12 + ' 2012').to_s] = dayname
  end
end

[holidays11, holidays12].each do |year|
  states = year.keys
  states.each do |state|
    year[state].each do |date, dayname,|
      if states.all? {|other_state| year[other_state].keys.any? {|other_date| other_date == date && year[other_state][other_date].upcase == dayname.upcase }}
        states.each {|s| year[s].delete(date) }
        year['National'][date] = dayname
      end
    end
  end
end
        
holidays = holidays11
holidays12.each do |state, dates|
  holidays[state].merge!(dates)
end

puts "{"
puts(holidays.map do |state, dates|
  "  '#{state}' => {\n" +
  dates.sort.map {|date, dayname| "    '#{date}' => '#{dayname}'" }.join(",\n") +
  "\n  }"
end.join(",\n"))
puts "}"


lib/oz_holidays.rb

require 'date'

class OzHolidays
  def self.national
    by_state['National']
  end
  def self.for_state(state)
    national.merge(by_state[state])
  end
  def self.[](state)
    for_state(state)
  end

  def self.to_json
    by_state.to_json
  end
  
  def self.by_date
    by(:date, :title)
  end

  def self.by(field1, field2)
    field3 = (all.first.keys - [field1, field2]).first
    result = Hash.new {|h,k| h[k] = {} }
    all.inject(Hash.new {|h,k| h[k] = {} }) do |result, hol|
      if result[hol[field1]].has_key? hol[field2]
        result[hol[field1]][hol[field2]] = [result[hol[field1]][hol[field2]], hol[field3]].flatten
      else
        result[hol[field1]][hol[field2]] = hol[field3]
      end
      result
    end
  end
  
  def self.by_state
    by(:state, :date)
  end

  def self.all
    [
      {:state => 'QLD',      :date => Date.new(2012,  5,  7), :title => 'Labour Day'             },
      {:state => 'QLD',      :date => Date.new(2012,  6, 11), :title => 'Queens Birthday'        },
      {:state => 'QLD',      :date => Date.new(2012,  8, 15), :title => 'Royal Qland Show metro' },
      {:state => 'QLD',      :date => Date.new(2013,  3, 30), :title => 'Easter Saturday'        },
      {:state => 'QLD',      :date => Date.new(2013,  5,  6), :title => 'Labour Day'             },
      {:state => 'QLD',      :date => Date.new(2013, 10,  7), :title => 'Queens Birthday'        },
      {:state => 'QLD',      :date => Date.new(2014,  6,  9), :title => 'Queens Birthday'        },
      {:state => 'QLD',      :date => Date.new(2014, 10,  6), :title => 'Labour Day'             },
      
      {:state => 'SA',       :date => Date.new(2012,  3, 12), :title => 'Adelaide Cup'           },
      {:state => 'SA',       :date => Date.new(2012,  6, 11), :title => 'Queens Birthday'        },
      {:state => 'SA',       :date => Date.new(2012, 10,  1), :title => 'Labour Day'             },
      {:state => 'SA',       :date => Date.new(2013,  3, 11), :title => 'Adelaide Cup'           },
      {:state => 'SA',       :date => Date.new(2013,  3, 30), :title => 'Easter Saturday'        },
      {:state => 'SA',       :date => Date.new(2013,  6, 10), :title => 'Queens Birthday'        },
      {:state => 'SA',       :date => Date.new(2013, 10,  7), :title => 'Labour Day'             },
      {:state => 'SA',       :date => Date.new(2014,  3, 10), :title => 'Adelaide Cup'           },
      {:state => 'SA',       :date => Date.new(2014,  6,  9), :title => 'Queens Birthday'        },
      {:state => 'SA',       :date => Date.new(2014, 10,  7), :title => 'Labour Day'             },
      
      {:state => 'NSW',      :date => Date.new(2012,  6, 11), :title => 'Queens Birthday'        },
      {:state => 'NSW',      :date => Date.new(2012, 10,  1), :title => 'Labour Day'             },
      {:state => 'NSW',      :date => Date.new(2013,  3, 30), :title => 'Easter Saturday'        },
      {:state => 'NSW',      :date => Date.new(2013,  6, 10), :title => 'Queens Birthday'        },
      {:state => 'NSW',      :date => Date.new(2013, 10,  7), :title => 'Labour Day'             },
      {:state => 'NSW',      :date => Date.new(2014,  6,  9), :title => 'Queens Birthday'        },
      {:state => 'NSW',      :date => Date.new(2014, 10,  6), :title => 'Labour Day'             },
      
      {:state => 'WA',       :date => Date.new(2012,  3,  5), :title => 'Labour Day'             },
      {:state => 'WA',       :date => Date.new(2012,  6,  4), :title => 'Foundation Day'         },
      {:state => 'WA',       :date => Date.new(2012, 10,  1), :title => 'Queens Birthday'        },
      {:state => 'WA',       :date => Date.new(2013,  3,  4), :title => 'Labour Day'             },
      {:state => 'WA',       :date => Date.new(2013,  6,  3), :title => 'Foundation Day'         },
      {:state => 'WA',       :date => Date.new(2013,  9, 30), :title => 'Queens Birthday'        },
      {:state => 'WA',       :date => Date.new(2014,  3,  3), :title => 'Labour Day'             },
      {:state => 'WA',       :date => Date.new(2014,  6,  2), :title => 'Foundation Day'         },
      {:state => 'WA',       :date => Date.new(2014,  9, 29), :title => 'Queens Birthday'        },


      {:state => 'VIC',      :date => Date.new(2012,  3, 12), :title => 'Labour Day'             },
      {:state => 'VIC',      :date => Date.new(2012,  6, 11), :title => 'Queens Birthday'        },
      {:state => 'VIC',      :date => Date.new(2012, 11,  6), :title => 'Melbourne Cup Day'      },
      {:state => 'VIC',      :date => Date.new(2013,  3, 11), :title => 'Labour Day'             },
      {:state => 'VIC',      :date => Date.new(2013,  3, 30), :title => 'Easter Saturday'        },
      {:state => 'VIC',      :date => Date.new(2013,  6, 10), :title => 'Queens Birthday'        },
      {:state => 'VIC',      :date => Date.new(2013, 11,  5), :title => 'Melbourne Cup Day'      },
      {:state => 'VIC',      :date => Date.new(2014,  3, 10), :title => 'Labour Day'             },
      {:state => 'VIC',      :date => Date.new(2014,  6,  9), :title => 'Queens Birthday'        },
      {:state => 'VIC',      :date => Date.new(2014, 11,  4), :title => 'Melbourne Cup Day'      },

      {:state => 'TAS',      :date => Date.new(2012,  2, 13), :title => 'Royal Hobart Regatta'   },
      {:state => 'TAS',      :date => Date.new(2012,  3, 12), :title => 'Eight Hours Day'        },
      {:state => 'TAS',      :date => Date.new(2012,  6, 11), :title => 'Queens Birthday'        },
      {:state => 'TAS',      :date => Date.new(2013,  3, 11), :title => 'Eight Hours Day'        },
      {:state => 'TAS',      :date => Date.new(2013,  6, 10), :title => 'Queens Birthday'        },
      {:state => 'TAS',      :date => Date.new(2014,  3, 10), :title => 'Eight Hours Day'        },
      {:state => 'TAS',      :date => Date.new(2014,  6,  9), :title => 'Queens Birthday'        },

      {:state => 'NT',       :date => Date.new(2012,  5,  7), :title => 'May Day'                },
      {:state => 'NT',       :date => Date.new(2012,  6, 11), :title => 'Queens Birthday'        },
      {:state => 'NT',       :date => Date.new(2012,  8,  6), :title => 'Picnic Day'             },
      {:state => 'NT',       :date => Date.new(2013,  3, 30), :title => 'Easter Saturday'        },
      {:state => 'NT',       :date => Date.new(2013,  6, 10), :title => 'Queens Birthday'        },
      {:state => 'NT',       :date => Date.new(2014,  5,  5), :title => 'May Day'                },
      {:state => 'NT',       :date => Date.new(2014,  8,  4), :title => 'Picnic Day'             },

      {:state => 'ACT',      :date => Date.new(2012,  3, 12), :title => 'Canberra Day'           },
      {:state => 'ACT',      :date => Date.new(2012,  6, 11), :title => 'Queens Birthday'        },
      {:state => 'ACT',      :date => Date.new(2012, 10,  1), :title => 'Labour Day'             },
      {:state => 'ACT',      :date => Date.new(2013,  3, 30), :title => 'Easter Saturday'        },
      {:state => 'ACT',      :date => Date.new(2013,  6, 10), :title => 'Queens Birthday'        },
      {:state => 'ACT',      :date => Date.new(2013, 10,  7), :title => 'Labour Day'             },
      {:state => 'ACT',      :date => Date.new(2014,  3, 10), :title => 'Canberra Day'           },
      {:state => 'ACT',      :date => Date.new(2014,  6,  9), :title => 'Queens Birthday'        },
      {:state => 'ACT',      :date => Date.new(2014,  9, 29), :title => 'Family Day'             },

      {:state => 'National', :date => Date.new(2012,  1,  2), :title => 'New Years Day'          },
      {:state => 'National', :date => Date.new(2012,  1, 26), :title => 'Australia Day'          },
      {:state => 'National', :date => Date.new(2012,  4,  6), :title => 'Good Friday'            },
      {:state => 'National', :date => Date.new(2012,  4,  9), :title => 'Easter Monday'          },
      {:state => 'National', :date => Date.new(2012,  4, 25), :title => 'ANZAC Day'              },
      {:state => 'National', :date => Date.new(2012, 12, 25), :title => 'Christmas Day'          },
      {:state => 'National', :date => Date.new(2012, 12, 26), :title => 'Boxing Day'             },
      {:state => 'National', :date => Date.new(2013,  1,  1), :title => 'New Years Day'          },
      {:state => 'National', :date => Date.new(2013,  1, 28), :title => 'Australia Day'          },
      {:state => 'National', :date => Date.new(2013,  3, 29), :title => 'Good Friday'            },
      {:state => 'National', :date => Date.new(2013,  4,  1), :title => 'Easter Monday'          },
      {:state => 'National', :date => Date.new(2013,  4, 25), :title => 'ANZAC Day'              },
      {:state => 'National', :date => Date.new(2013, 12, 25), :title => 'Christmas Day'          },
      {:state => 'National', :date => Date.new(2013, 12, 26), :title => 'Boxing Day'             },
      {:state => 'National', :date => Date.new(2014,  1,  1), :title => 'New Years Day'          },
      {:state => 'National', :date => Date.new(2014,  1, 27), :title => 'Australia Day'          },
      {:state => 'National', :date => Date.new(2014,  4, 18), :title => 'Good Friday'            },
      {:state => 'National', :date => Date.new(2014,  4, 21), :title => 'Easter Monday'          },
      {:state => 'National', :date => Date.new(2014,  4, 25), :title => 'ANZAC Day'              },
      {:state => 'National', :date => Date.new(2014, 12, 25), :title => 'Christmas Day'          },
      {:state => 'National', :date => Date.new(2014, 12, 26), :title => 'Boxing Day'             },


    ]
  end
end


javascript/oz_holidays.js
var oz_holidays = {
  "QLD":{"2012-05-07":"Labour Day","2012-06-11":"Queens Birthday","2012-08-15":"Royal Qland Show metro","2013-03-30":"Easter Saturday","2013-05-06":"Labour Day","2013-10-07":"Queens Birthday","2014-06-09":"Queens Birthday","2014-10-06":"Labour Day"},
  "SA":{"2012-03-12":"Adelaide Cup","2012-06-11":"Queens Birthday","2012-10-01":"Labour Day","2013-03-11":"Adelaide Cup","2013-03-30":"Easter Saturday","2013-06-10":"Queens Birthday","2013-10-07":"Labour Day","2014-03-10":"Adelaide Cup","2014-06-09":"Queens Birthday","2014-10-07":"Labour Day"},
  "NSW":{"2012-06-11":"Queens Birthday","2012-10-01":"Labour Day","2013-03-30":"Easter Saturday","2013-06-10":"Queens Birthday","2013-10-07":"Labour Day","2014-06-09":"Queens Birthday","2014-10-06":"Labour Day"},
  "WA":{"2012-03-05":"Labour Day","2012-06-04":"Foundation Day","2012-10-01":"Queens Birthday","2013-03-04":"Labour Day","2013-06-03":"Foundation Day","2013-09-30":"Queens Birthday","2014-03-03":"Labour Day","2014-06-02":"Foundation Day","2014-09-29":"Queens Birthday"},
  "VIC":{"2012-03-12":"Labour Day","2012-06-11":"Queens Birthday","2012-11-06":"Melbourne Cup Day","2013-03-11":"Labour Day","2013-03-30":"Easter Saturday","2013-06-10":"Queens Birthday","2013-11-05":"Melbourne Cup Day","2014-03-10":"Labour Day","2014-06-09":"Queens Birthday","2014-11-04":"Melbourne Cup Day"},
  "TAS":{"2012-02-13":"Royal Hobart Regatta","2012-03-12":"Eight Hours Day","2012-06-11":"Queens Birthday","2013-03-11":"Eight Hours Day","2013-06-10":"Queens Birthday","2014-03-10":"Eight Hours Day","2014-06-09":"Queens Birthday"},
  "NT":{"2012-05-07":"May Day","2012-06-11":"Queens Birthday","2012-08-06":"Picnic Day","2013-03-30":"Easter Saturday","2013-06-10":"Queens Birthday","2014-05-05":"May Day","2014-08-04":"Picnic Day"},
  "ACT":{"2012-03-12":"Canberra Day","2012-06-11":"Queens Birthday","2012-10-01":"Labour Day","2013-03-30":"Easter Saturday","2013-06-10":"Queens Birthday","2013-10-07":"Labour Day","2014-03-10":"Canberra Day","2014-06-09":"Queens Birthday","2014-09-29":"Family Day"},
  "National":{"2012-01-02":"New Years Day","2012-01-26":"Australia Day","2012-04-06":"Good Friday","2012-04-09":"Easter Monday","2012-04-25":"ANZAC Day","2012-12-25":"Christmas Day","2012-12-26":"Boxing Day","2013-01-01":"New Years Day","2013-01-28":"Australia Day","2013-03-29":"Good Friday","2013-04-01":"Easter Monday","2013-04-25":"ANZAC Day","2013-12-25":"Christmas Day","2013-12-26":"Boxing Day","2014-01-01":"New Years Day","2014-01-27":"Australia Day","2014-04-18":"Good Friday","2014-04-21":"Easter Monday","2014-04-25":"ANZAC Day","2014-12-25":"Christmas Day","2014-12-26":"Boxing Day"}
}


gen/output.rb
{
  'QLD' => {
    '2011-04-25' => 'ANZAC Day',
    '2011-04-26' => 'Easter Monday',
    '2011-05-02' => 'Labour Day',
    '2011-06-13' => 'Queens Birthday',
    '2011-08-17' => 'Royal Qland Show metro',
    '2012-05-07' => 'Labour Day',
    '2012-06-11' => 'Queens Birthday',
    '2012-08-15' => 'Royal Qland Show metro'
  },
  'SA' => {
    '2011-03-14' => 'Adelaide Cup',
    '2011-04-25' => 'Easter Monday',
    '2011-04-26' => 'ANZAC Day',
    '2011-06-13' => 'Queens Birthday',
    '2011-10-03' => 'Labour Day',
    '2012-03-12' => 'Adelaide Cup',
    '2012-06-11' => 'Queens Birthday',
    '2012-10-01' => 'Labour Day'
  },
  'NSW' => {
    '2011-04-25' => 'ANZAC Day',
    '2011-04-26' => 'Easter Monday',
    '2011-06-13' => 'Queens Birthday',
    '2011-10-03' => 'Labour Day',
    '2012-06-11' => 'Queens Birthday',
    '2012-10-01' => 'Labour Day'
  },
  'WA' => {
    '2011-03-07' => 'Labour Day',
    '2011-04-25' => 'Easter Monday',
    '2011-04-26' => 'ANZAC Day',
    '2011-06-06' => 'Foundation Day',
    '2011-10-03' => 'Queens Birthday',
    '2012-03-05' => 'Labour Day',
    '2012-06-04' => 'Foundation Day',
    '2012-10-01' => 'Queens Birthday'
  },
  'VIC' => {
    '2011-03-14' => 'Labour Day',
    '2011-04-25' => 'Easter Monday',
    '2011-04-26' => 'ANZAC Day',
    '2011-06-13' => 'Queens Birthday',
    '2011-11-01' => 'Melbourne Cup Day',
    '2012-03-12' => 'Labour Day',
    '2012-06-11' => 'Queens Birthday',
    '2012-11-06' => 'Melbourne Cup Day'
  },
  'TAS' => {
    '2011-02-14' => 'Royal Hobart Regatta',
    '2011-03-14' => 'Eight Hours Day',
    '2011-04-25' => 'ANZAC Day',
    '2011-06-13' => 'Queens Birthday',
    '2012-02-13' => 'Royal Hobart Regatta',
    '2012-03-12' => 'Eight Hours Day',
    '2012-06-11' => 'Queens Birthday'
  },
  'NT' => {
    '2011-04-25' => 'ANZAC Day',
    '2011-05-02' => 'May Day',
    '2011-06-13' => 'Queens Birthday',
    '2011-08-01' => 'Picnic Day',
    '2012-05-07' => 'May Day',
    '2012-06-11' => 'Queens Birthday',
    '2012-08-06' => 'Picnic Day'
  },
  'ACT' => {
    '2011-03-14' => 'Canberra Day',
    '2011-04-25' => 'Easter Monday',
    '2011-04-26' => 'ANZAC Day',
    '2011-06-13' => 'Queens Birthday',
    '2011-10-03' => 'Labour Day',
    '2012-03-12' => 'Canberra Day',
    '2012-06-11' => 'Queens Birthday',
    '2012-10-01' => 'Labour Day'
  },
  'National' => {
    '2011-01-03' => 'New Years Day',
    '2011-01-26' => 'Australia Day',
    '2011-04-22' => 'Good Friday',
    '2011-12-26' => 'Christmas Day',
    '2011-12-27' => 'Boxing Day',
    '2012-01-02' => 'New Years Day',
    '2012-01-26' => 'Australia Day',
    '2012-04-06' => 'Good Friday',
    '2012-04-09' => 'Easter Monday',
    '2012-04-25' => 'ANZAC Day',
    '2012-12-25' => 'Christmas Day',
    '2012-12-26' => 'Boxing Day'
  }
}
